keps/sig-network/2732-np-version/README.md

```diff
@@ -0,0 +1,803 @@
+
+As a network plugin developer, I want to verify that a CNI implements a specific
+NetworkPolicy version.
+
+#### Story 5 - Changing an Alpha NetworkPolicy API
+
+As a kubernetes developer, I want to make a change to a previously-added
+alpha NetworkPolicy feature, in a way that does not cause problems for
+users in clusters with varying apiserver and network plugin versions.
+
+### Notes/Constraints/Caveats
+
+#### Feature Gates and NetworkPolicy
+
+Alpha APIs create an additional problem for NetworkPolicy, due to the
+three-way (user/apiserver/plugin) versioning split; if an alpha field
+is added in 1.21 but then changed in 1.22, then a 1.22 cluster with a
+network plugin implementing the 1.21 version would probably be unable
+to use either version of the feature. We might want to establish a
+somewhat more restrictive set of alpha API rules for NetworkPolicy
+changes than are used for other API types.
+
+#### Determining Whether a Network Plugin Implements NetworkPolicyStatus
+
+If a user is waiting to see the status of a newly-created NetworkPolicy,
+there is no entirely-reliable way to distinguish "the plugin has not yet
+set `status` but will soon" from "the plugin doesn't know about `status`
+and is never going to set it".
+
+It's not clear how big a problem this is, especially if we suggest that
+implementations should create an "empty" `status` right away if it's
+going to take them a while to determine the final `status`.
+
+#### Distributed and Delegating NetworkPolicy Implementations
+
+Some NetworkPolicy implementations do not operate from a single
+controller, but instead operate independently on every node (much like
+how kube-proxy implements Service rules). In this case it is not clear
+who would be responsible for updating `NetworkPolicyStatus`, and it may
+not be easy for any part of the system to know with certainty that a
+given NetworkPolicy was fully in effect across the cluster.
+
+In other cases, a plugin may implement NetworkPolicy in terms of some
+other lower-level policy engine (implemented by a cloud, perhaps), and
+may work by simply translating NetworkPolicy rules into rules for this
+other engine. As in the distributed case, it may not be possible for
+such a NetworkPolicy implementation to know for sure when a policy was
+fully in effect.
+
+### Risks and Mitigations
+
+## Design Details
+
+### API
+
+Add a new field to `networkingv1.NetworkPolicySpec`:
+
+```
+type NetworkPolicySpec struct {
+	...
+
+	// minVersion indicates the Kubernetes release version corresponding to
+	// the NetworkPolicy functionality required by this policy. If this is
+	// explicitly specified by the user, the apiserver will require that the policy
+	// conforms to the specified version. If it is not specified, the apiserver
+	// will fill in the correct minVersion based on the features used by the policy.
+	// +optional
+	MinVersion NetworkPolicyVersion `json:"minVersion,omitempty" protobuf:"bytes,5,name=minVersion"`
```

Comment:
This implies that version is a linear feature-accumulator.  I'm not convinced that is 100% accurate.

Suppose we start with version X.  Then version Y adds "foo" support.  Then version Z adds "bar" support.

If an implementation supports "bar' and not "foo" - is that allowed?  At leas you can assert that if it knows about "bar" it *should* be aware of "foo".  But are we intending to mandate 100% implementation?  How can we enforce that?

An alternative might be a list of feature flags.  E.g. "this policy uses: [ FeatureFoo, FeatureBar ]".  Then controllers can warn "I support Bar, but not Foo".  Of course, we have to get all policy imps to support the features-required field...

Reply:
I'd also say this is NOT something a user may set themselves.

Reply:
> Suppose we start with version X. Then version Y adds "foo" support. Then version Z adds "bar" support.
> 
> If an implementation supports "bar' and not "foo" - is that allowed? At least you can assert that if it knows about "bar" it should be aware of "foo". But are we intending to mandate 100% implementation? How can we enforce that?

The intention in my original KEP is that a network policy implementation has to _recognize_ every feature up to the latest feature that it implements. So an implementation can support "bar" and not "foo", but it would be required to recognize when a policy was using "foo", and fail in an appropriate way. This is discussed more later, in the `NetworkPolicyVersion` descriptions and the section "The Supported Condition"